Speed up `replace` and `canonicalize` by improving type stability

SebastianM-C opened the pull request:
## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

This PR attempts to fix #2608 by avoiding boxing and runtime dispatch. So far this works for the case of homogeneous parameters, but for heterogeneous ones, we end up iterating over a tuple of  different types, which is unstable.

AayushSabharwal commented on src/systems/parameter_buffer.jl:
```diff
@@ -171,7 +187,7 @@ for (Portion, field) in [(SciMLStructures.Tunable, :tunable)
             function (new_val)
                 if new_val !== as_vector
-                    update_tuple_of_buffers(new_val, p.$field)
+                    @set! p.$field = update_tuple_of_buffers(new_val, p.$field)
```
`update_tuple_of_buffers` was written to be an in-place update. If we want to do an out-of-place update here, we should use `split_into_buffers` and delete `update_tuple_of_buffers`

SebastianM-C replied:
Ah, that might explain the test failures 😅

SebastianM-C replied:
I think one question is whether `repack` is supposed to be in-place or out-of-place. I'm not very clear on that.

AayushSabharwal replied:
How I understand it:
Since `alias = true`, if `repack` receives the same array object that `canonicalize` returned, it doesn't need to do anything. Since `MTKParameters` is a mutable SciMLStructure, `repack` should ideally be in-place.

SebastianM-C replied:
But if I want to promote the element type of a `$field`, how would I use `replace`? For `replace` I'd have to use the `newvals`, but that would need the canonicalized version of the parameters, which is why I was thinking to use `repack`.

AayushSabharwal replied:
You don't _have_ to use `repack` if you called `canonicalize`, could just use `replace`.

SebastianM-C commented:
I fixed a few more cases and added tests with JET and `@inferred` for `replace`/`replace!`. For `canonicalize` there are some cases which can't be inferred due to storing things in vectors, but that could potentially be addressed in the future with a  vector type that stores its size in the type domain. I was a bit surprised that `replace` avoids the issue, but I think that the main factor is that we're effectively unrolling the iterations at compile time.

I think that this should not increase compile times as the tuples are used for the number of types, so we don't expect a ridiculous number of types in the portions.

AayushSabharwal reviewed (approved):
LGTM!
